### PR TITLE
Don't extend base functions on base types

### PR DESCRIPTION
**Pull request by tkelman**
Please don't monkey-patch Base, that has global side effects

### CODE REVIEW

**Comment by tkelman** [comment]
@MikeInnes hello?

File: test/recurrent.jl [hunk 1]
```diff
@@ -13,5 +13,5 @@ end
   _, ys = apply(unroll1(r).model, xs, (r.y.x,))
   @test ys[1] == tanh(xs[1] * r.Wxy.x .+ r.y.x * r.Wyy.x .+ r.by.x)
   ru = unroll(r, 3)
-  ru(batchone(Seq(squeeze.(xs))))[1] == squeeze.(ys)
+  ru(batchone(Seq((x->squeeze(x,1)).(xs))))[1] == (x->squeeze(x,1)).(ys)
```
Comment (MikeInnes) [inline test/recurrent.jl]: This should be `squeeze.(xs, 1)`

Comment (tkelman) [inline test/recurrent.jl]: of course, not sure why I didn't use that first

Comment (tkelman) [inline test/recurrent.jl]: done

**Comment by MikeInnes** [comment]
I think a nicer solution would be to add this method to Base, to be honest.

**Comment by tkelman** [comment]
That should be done via a PR to base (then backported via Compat once merged), not potentially concealing bugs in unrelated code in some package. And I don't think it makes much sense given the docstring is
```
  squeeze(A, dims)

  Remove the dimensions specified by dims from array A. Elements of dims must be unique and within the range
  1:ndims(A). size(A,i) must equal 1 for all i in dims.
```

Unless there are many of these dimension arguments that default to 1 in other functions that I'm not thinking of.

**Comment by tkelman** [comment]
Could give this a local `squeeze1` name for the time being, or `Flux.squeeze` and not import it, to avoid side effects. Whatever you prefer that can avoid the type piracy.

**Comment by tkelman** [comment]
Can we move on this? Changing the behavior of totally unrelated code depending on whether or not this package is imported is not okay.

**Comment by tkelman** [comment]
Bump again, this would be incompatible with the proposed meaning in https://github.com/JuliaLang/julia/pull/22000, all the more reason it shouldn't be extended here.

**Comment by MikeInnes** [comment]
Perhaps it would be better for Flux to have its own, unexported `squeeze` with these semantics so it can be imported by users.

**Comment by tkelman** [comment]
Sure, that works too, done.

**Comment by staticfloat** [comment]
@MikeInnes fwiw, a bunch of the discussion in the above linked issue basically boiled down to me realizing that for my use cases, what I really wanted was to just use `vec()`, which takes in things like Nx1 matrices and spits out N length vectors.  It's a little weird in that it can also take in NxN matrices and spit out (N^2) vectors, but in your use case that might not be a problem.

**Comment by MikeInnes** [comment]
In this case it comes up because the first dim is a batch – but often you just want to work with one (N-d) sample, so it's convenient to wrap and unwrap that in a batch of size 1. So `batch` and `unbatch` do map to `[un]squeeze(xs, 1)` and there's no real way to get around that.

Of course, it'd be much nicer to have batching be first class in the type system rather than representing it implicitly like this, but that's another project.

**Pull request merged**